### PR TITLE
executor, infoschema: `SHOW COLLATIONS` shows supported collations only

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve? 

This PR aims to fix 2 compatibility issues:
 
* In TIDB, the `SHOW CHARSET` command shows only supported charsets:
```
mysql> show charset;
+---------+---------------+-------------------+--------+
| Charset | Description   | Default collation | Maxlen |
+---------+---------------+-------------------+--------+
| utf8    | UTF-8 Unicode | utf8_bin          |      3 |
| utf8mb4 | UTF-8 Unicode | utf8mb4_bin       |      4 |
| ascii   | US ASCII      | ascii_bin         |      1 |
| latin1  | Latin1        | latin1_bin        |      1 |
| binary  | binary        | binary            |      1 |
+---------+---------------+-------------------+--------+
``` 
But `SHOW COLLATIONS` shows all collations that supported in MySQL 5.7:
```
mysql> show collation;
+--------------------------+----------+------+---------+----------+---------+
| Collation                | Charset  | Id   | Default | Compiled | Sortlen |
+--------------------------+----------+------+---------+----------+---------+
| big5_chinese_ci          | big5     |    1 | Yes     | Yes      |       1 |
| latin2_czech_cs          | latin2   |    2 |         | Yes      |       1 |
| dec8_swedish_ci          | dec8     |    3 | Yes     | Yes      |       1 |
| cp850_general_ci         | cp850    |    4 | Yes     | Yes      |       1 |
| latin1_german1_ci        | latin1   |    5 |         | Yes      |       1 |
| hp8_english_ci           | hp8      |    6 | Yes     | Yes      |       1 |
| koi8r_general_ci         | koi8r    |    7 | Yes     | Yes      |       1 |
| latin1_swedish_ci        | latin1   |    8 | Yes     | Yes      |       1 |
| latin2_general_ci        | latin2   |    9 | Yes     | Yes      |       1 |
...
| utf8mb4_german2_ci       | utf8mb4  |  244 |         | Yes      |       1 |
| utf8mb4_croatian_ci      | utf8mb4  |  245 |         | Yes      |       1 |
| utf8mb4_unicode_520_ci   | utf8mb4  |  246 |         | Yes      |       1 |
| utf8mb4_vietnamese_ci    | utf8mb4  |  247 |         | Yes      |       1 |
+--------------------------+----------+------+---------+----------+---------+
219 rows in set (0.01 sec)
```
It's better for `SHOW COLLATIONS` to display only supported collations.

* The `Default` column is wrong for some of `SHOW COLLATION` results:
```
mysql> show collation where Collation='utf8_bin';
+-----------+---------+------+---------+----------+---------+
| Collation | Charset | Id   | Default | Compiled | Sortlen |
+-----------+---------+------+---------+----------+---------+
| utf8_bin  | utf8    |   83 |         | Yes      |       1 |
+-----------+---------+------+---------+----------+---------+
1 row in set (0.00 sec)
```
The default collation of `utf8` in TiDB is `utf8_bin`

### What is changed and how it works?
This main change is from parser PR: https://github.com/pingcap/parser/pull/295

### Check List
Tests 
 - Integration test

Code changes
 - Has interface methods change

Side effects
 - Breaking backward compatibility in some `SHOW` result

Related changes
 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by morgo** [comment]
Thank you for working on this. I think it's a great enhancement!

File: executor/show.go [hunk 1]
```diff
@@ -861,7 +861,7 @@ func (e *ShowExec) fetchShowCreateDatabase() error {
 }
 
 func (e *ShowExec) fetchShowCollation() error {
-	collations := charset.GetCollations()
```
Comment (crazycs520) [inline executor/show.go]: can we remove this function in parser now?

**Comment by tiancaiamao** [comment]
LGTM

**Comment by zz-jason** [comment]
@bb7133 Please merge master && resolve conflicts. BTW please add some proper labels for this PR.

**Comment by tiancaiamao** [comment]
The circle CI fail is misleading, it actually pass all tests
I remove the redundant 'parser replace check'
https://github.com/pingcap/tidb/pull/10228

File: executor/show.go (hunk 1 above)
Comment (bb7133) [reply executor/show.go]: It looks reasonable, I will try to remove it in another PR, thanks

**Comment by bb7133** [comment]
PTAL @crazycs520 @zz-jason

**Pull request merged**